qiita_db/sql_connection.py

```diff
@@ -545,99 +436,156 @@ def fetchall_with_types(self, sql, sql_args=None):
+
+    def add_to_queue(self, queue, sql, sql_args=None, many=False):
+        """Add an sql command to the end of a queue
 
         Parameters
         ----------
+        queue : str
+            name of queue adding to
         sql : str
-            The SQL query
-        sql_args : tuple or list, optional
-            The arguments for the SQL query
+            sql command to run
+        sql_args : list, tuple or dict, optional
+            the arguments to fill sql command with
+        many : bool, optional
+            Whether or not this should be treated as an executemany command.
+            Default False
 
         Raises
         ------
-        QiitaDBExecutionError
-            if there is some error executing the SQL query
+        KeyError
+            queue does not exist
 
-        Note: from psycopg2 documentation, only variable values should be bound
-            via sql_args, it shouldn't be used to set table or field names. For
-            those elements, ordinary string formatting should be used before
-            running execute.
+        Notes
+        -----
+        Queues are executed in FIFO order
         """
-        with self._sql_executor(sql, sql_args):
-            pass
+        if not many:
+            sql_args = [sql_args]
 
-    def executemany(self, sql, sql_args_list):
-        """ Executes an executemany SQL query with no results
+        for args in sql_args:
+            self._check_sql_args(args)
+            self.queues[queue].append((sql, args))
+
+    def _rollback_raise_error(self, queue, sql, sql_args, e):
+        self._connection.rollback()
+        # wipe out queue since it has an error in it
+        del self.queues[queue]
+        raise QiitaDBExecutionError(
+            ("\nError running SQL query in queue %s: %s\nARGS: %s\nError: %s"
+             % (queue, sql, str(sql_args), e)))
+
+    def execute_queue(self, queue):
+        """Executes all sql in a queue in a single transaction block
 
         Parameters
         ----------
-        sql : str
-            The SQL query
-        sql_args : list of tuples
-            The arguments for the SQL query
+        queue : str
+            Name of queue to execute
 
-        Raises
-        ------
-        QiitaDBExecutionError
-            If there is some error executing the SQL query
+        Notes
+        -----
+        Does not support executemany command. Instead, enter the multiple
+        SQL commands as multiple entries in the queue.
 
-        Note: from psycopg2 documentation, only variable values should be bound
-            via sql_args, it shouldn't be used to set table or field names. For
-            those elements, ordinary string formatting should be used before
-            running execute.
+        Queues are executed in FIFO order
         """
-        with self._sql_executor(sql, sql_args_list, True):
-            pass
+        self._check_queue_exists(queue)
+
+        with self.get_postgres_cursor() as cur:
```

Comment:
Yeah... I will try to decompose this. When even its own original developer doesn't know what is going on in the code (@squirrelo) means that the code is really convoluted. Note that I just moved this function around :smile: